seed/php-sdk/imdb/src/Core/JsonSerializer.php

```diff
@@ -0,0 +1,137 @@
+<?php
+
+namespace Seed\Core;
+
+use DateTime;
+use Exception;
+use JsonException;
+
+class JsonSerializer
+{
+    public static function serializeDate(DateTime $date): string {
+        return $date->format(Constant::DateFormat);
+    }
+
+    public static function serializeDateTime(DateTime $date): string {
+        return $date->format(Constant::DateTimeFormat);
+    }
+
+    /**
+     * Serializes the given array based on the type annotation.
+     *
+     * @param mixed[]|array<string, mixed> $data The array to be serialized.
+     * @param mixed[]|array<string, mixed> $type The type definition from the annotation.
+     * @return mixed[]|array<string, mixed> The serialized array.
+     * @throws JsonException
+     */
+    public static function serializeArray(array $data, array $type): array
+    {
+        return Utils::isMapType($type)
+            ? self::serializeMap($data, $type)
+            : self::serializeList($data, $type);
+    }
+
+    /**
+     * Serializes a value based on the type definition.
+     *
+     * @param mixed $data The value to serialize.
+     * @param mixed $type The type definition.
+     * @return mixed The serialized value.
+     * @throws JsonException
+     */
+    private static function serializeValue(mixed $data, mixed $type): mixed
+    {
+        if ($type instanceof Union) {
+            foreach ($type->types as $unionType) {
+                try {
+                    return self::serializeSingleValue($data, $unionType);
+                } catch (Exception $e) {
+                    continue;
+                }
+            }
+            throw new \InvalidArgumentException("Cannot serialize value with any of the union types.");
+        }
+        if (is_array($type)) {
+            return self::serializeArray((array)$data, $type);
+        }
+        if (gettype($type) != "string") {
+            throw new JsonException("Unexpected non-string type.");
+        }
+        return self::serializeSingleValue($data, $type);
+    }
+
+    /**
+     * Serializes a single value based on its type.
+     *
+     * @param mixed $data The value to serialize.
+     * @param string $type The expected type.
+     * @return mixed The serialized value.
+     */
+    private static function serializeSingleValue(mixed $data, string $type): mixed
+    {
+        if ($type === 'null' && $data === null) {
+            return null;
+        }
+
+        if (($type === 'date' || $type === 'datetime') && $data instanceof DateTime) {
+            return $type === 'date' ? self::serializeDate($data) : self::serializeDateTime($data);
+        }
```

Comment:
What's the behavior if `$data` is a `DateTime` but `$type` doesn't match `date` or `datetime`? Should we throw an error?

Reply:
we'd fall through to the throw at the end of the method